--- a/src/libcharon/plugins/kernel_netlink/kernel_netlink_net.c
+++ b/src/libcharon/plugins/kernel_netlink/kernel_netlink_net.c
@@ -2666,7 +2666,7 @@
 {
 	if (chunk_equals(a->route.dst_net, b->dst_net) &&
 		a->route.prefixlen == b->prefixlen &&
-		is_known_vip(a->this, b->src_ip))
+		b->src_ip && is_known_vip(a->this, b->src_ip))
 	{
 		return TRUE;
 	}
@@ -2713,13 +2713,15 @@
 
 	/* don't replace the route if we already have one with a VIP installed,
 	 * but keep track of it in case that other route is uninstalled */
-	this->lock->read_lock(this->lock);
-	if (!is_known_vip(this, src_ip))
-	{
-		found = this->routes->get_match(this->routes, &lookup,
-										(void*)route_with_vip);
+	if (src_ip) {
+		this->lock->read_lock(this->lock);
+		if (src_ip && !is_known_vip(this, src_ip))
+		{
+			found = this->routes->get_match(this->routes, &lookup,
+							(void*)route_with_vip);
+		}
+		this->lock->unlock(this->lock);
 	}
-	this->lock->unlock(this->lock);
 	if (found)
 	{
 		status = SUCCESS;
