--- a/src/libcharon/kernel/kernel_ipsec.h
+++ b/src/libcharon/kernel/kernel_ipsec.h
@@ -142,6 +142,8 @@
  * Data required to delete an SA in the kernel
  */
 struct kernel_ipsec_del_sa_t {
+	/** Reqid */
+	uint32_t reqid;
 	/** CPI in case IPComp is used */
 	uint16_t cpi;
 };
--- a/src/libcharon/sa/child_sa.c
+++ b/src/libcharon/sa/child_sa.c
@@ -1475,6 +1475,7 @@
 		.domain = this->config->get_domain(this->config),
 	};
 	kernel_ipsec_del_sa_t sa = {
+		.reqid = this->reqid,
 		.cpi = this->other_cpi,
 	};
 	charon->kernel->del_sa(charon->kernel, &id, &sa);
@@ -1758,6 +1759,7 @@
 			.domain = this->config->get_domain(this->config),
 		};
 		kernel_ipsec_del_sa_t sa = {
+			.reqid = this->reqid,
 			.cpi = this->my_cpi,
 		};
 		charon->kernel->del_sa(charon->kernel, &id, &sa);
@@ -1774,6 +1776,7 @@
 			.domain = this->config->get_domain(this->config),
 		};
 		kernel_ipsec_del_sa_t sa = {
+			.reqid = this->reqid,
 			.cpi = this->other_cpi,
 		};
 		charon->kernel->del_sa(charon->kernel, &id, &sa);
