--- a/src/libcharon/plugins/updown/updown_listener.c
+++ b/src/libcharon/plugins/updown/updown_listener.c
@@ -370,6 +370,14 @@
 	{
 		push_env(envp, countof(envp), "PLUTO_HOST_ACCESS=1");
 	}
+	if (config->get_domain(config))
+	{
+		push_env(envp, countof(envp), "PLUTO_DOMAIN=%s", config->get_domain(config));
+	}
+	if (config->get_domain(config))
+	{
+		push_env(envp, countof(envp), "PLUTO_CFG_INTERFACE=%s", config->get_interface(config));
+	}
 
 	process = process_start_shell(envp, NULL, &out, NULL, "2>&1 %s",
 								  config->get_updown(config));
