--- a/src/libcharon/plugins/kernel_netlink/kernel_netlink_ipsec.c
+++ b/src/libcharon/plugins/kernel_netlink/kernel_netlink_ipsec.c
@@ -2765,13 +2765,11 @@
 	/* install a route, if:
 	 * - this is an outbound policy (to just get one for each child)
 	 * - routing is not disabled via strongswan.conf
-	 * - the selector is not for a specific protocol/port
 	 * - no XFRM interface ID is configured
 	 * - we are in tunnel/BEET mode or install a bypass policy
 	 */
 	if (policy->direction == POLICY_OUT && this->install_routes &&
 		(policy->sel.prefixlen_s != 0 || ipsec->mark.value == 0) &&
-		!policy->sel.proto && !policy->sel.dport && !policy->sel.sport &&
 		!policy->if_id)
 	{
 		if (mapping->type == POLICY_PASS ||
